agent/xds/clusters.go

```diff
@@ -1828,12 +1830,15 @@ func configureClusterWithHostnames(
 			continue
 		}
 
-		if len(endpoints) == 0 {
+		if len(endpoints) < envoyMaxEndpoints {
 			endpoints = append(endpoints, makeLbEndpoint(addr, port, health, weight))
 
 			hostname = addr
 			idx = i
-			break
+
+			if len(endpoints) == envoyMaxEndpoints {
+				break
```

Comment:
Pretty sure we do still need to break early here to maintain the same behaviour that we had before, but happy to be corrected so we can remove this slightly clunky check.